src/backend/distributed/utils/foreign_key_relationship.c

```diff
@@ -98,6 +104,108 @@ static void GetConnectedListHelper(ForeignConstraintRelationshipNode *node,
 static List * GetForeignConstraintRelationshipHelper(Oid relationId, bool isReferencing);
 
 
+/*
+ * GetForeignKeyConnectedRelationIdList returns a list of relation id's for
+ * relations that are connected to relation with relationId via a foreign
+ * key graph.
+ */
+List *
+GetForeignKeyConnectedRelationIdList(Oid relationId)
+{
+	/* use ShareRowExclusiveLock to prevent concurent foreign key creation */
+	LOCKMODE lockMode = ShareRowExclusiveLock;
+	Relation relation = try_relation_open(relationId, lockMode);
+	if (!RelationIsValid(relation))
+	{
+		ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+						errmsg("relation with OID %d does not exist",
+							   relationId)));
+	}
+
+	relation_close(relation, NoLock);
+
+	bool foundInFKeyGraph = false;
+	ForeignConstraintRelationshipNode *relationshipNode =
+		GetRelationshipNodeForRelationId(relationId, &foundInFKeyGraph);
+	if (!foundInFKeyGraph)
+	{
+		/*
+		 * Relation could not be found in foreign key graph, then it has no
+		 * foreign key relationships.
+		 */
+		return NIL;
+	}
+
+	List *fKeyConnectedRelationshipNodeList =
+		GetRelationshipNodesForFKeyConnectedRelations(relationshipNode);
+	List *fKeyConnectedRelationIdList = GetRelationIdsFromRelationshipNodeList(
+		fKeyConnectedRelationshipNodeList);
+	return fKeyConnectedRelationIdList;
+}
+
+
+/*
+ * GetRelationshipNodesForFKeyConnectedRelations performs depth-first search
+ * starting from input ForeignConstraintRelationshipNode and returns a list
+ * of ForeignConstraintRelationshipNode objects for relations that are connected
+ * to given relation node via a foreign key relationhip graph.
+ */
+static List *
+GetRelationshipNodesForFKeyConnectedRelations(
+	ForeignConstraintRelationshipNode *relationshipNode)
+{
+	relationshipNode->visited = true;
+	List *relationshipNodeList = list_make1(relationshipNode);
+
+	ForeignConstraintRelationshipNode *currentNode = NULL;
+	foreach_ptr_append(currentNode, relationshipNodeList)
+	{
+		List *allNeighboursList = GetAllNeighboursList(currentNode);
+		ForeignConstraintRelationshipNode *neighbourNode = NULL;
+		foreach_ptr(neighbourNode, allNeighboursList)
+		{
+			if (neighbourNode->visited)
+			{
+				continue;
+			}
+
+			neighbourNode->visited = true;
```

Comment:
Can we do this without modifying global state?

Reply:
It's bad practice but seems that we use this field for such purposes by design:
https://github.com/citusdata/citus/blob/3f60b08b11c8b3ec5de14b737e78b6b29be67d0d/src/backend/distributed/utils/foreign_key_relationship.c#L53..L54

We could actually remove this `visited` flag from global struct and we could maintain `visited` flags in a seperate hash table that is local to function that performs dfs. If it makes sense, then I can open a seperate pr to remove this field first, then I can apply changes in this pr too.

What do you think @marcocitus  ?

Reply:
That would be good. We probably don't visit this logic very often, but given past experience it's likely to result in crashes / other issues at some point.

Reply:
opened a pr for that https://github.com/citusdata/citus/pull/4446